src/sources/http_scrape/scrape.rs

```diff
@@ -0,0 +1,243 @@
+    async fn build(&self, cx: SourceContext) -> Result<sources::Source> {
+        // build the url
+        let endpoints = vec![self.endpoint.clone()];
+        let urls = endpoints
+            .iter()
+            .map(|s| s.parse::<Uri>().context(sources::UriParseSnafu))
+            .map(|r| r.map(|uri| super::get_url(&uri, &self.query)))
+            .collect::<std::result::Result<Vec<Uri>, sources::BuildError>>()?;
+
+        let tls = TlsSettings::from_options(&self.tls)?;
+
+        // build the decoder
+        let decoder = DecodingConfig::new(
+            self.framing.clone(),
+            self.decoding.clone(),
+            LogNamespace::Vector,
+        )
+        .build();
+
+        let content_type = self.decoding.content_type(&self.framing).to_string();
+
+        // the only specific context needed is the codec decoding
+        let context = HttpScrapeContext { decoder };
+
+        let inputs = super::GenericHttpScrapeInputs {
+            urls,
+            interval_secs: self.scrape_interval_secs,
+            headers: self.headers.clone(),
+            content_type,
+            auth: self.auth.clone(),
+            tls,
+            proxy: cx.proxy.clone(),
+            shutdown: cx.shutdown,
+        };
+
+        Ok(super::http_scrape(inputs, context, cx.out).boxed())
+    }
+
+    fn outputs(&self, _global_log_namespace: LogNamespace) -> Vec<Output> {
+        vec![Output::default(self.decoding.output_type())]
+    }
+
+    fn source_type(&self) -> &'static str {
+        NAME
+    }
+
+    fn can_acknowledge(&self) -> bool {
+        false
+    }
+}
+
+#[derive(Clone)]
+struct HttpScrapeContext {
+    decoder: Decoder,
+}
+
+impl HttpScrapeContext {
+    /// Decode the events from the byte buffer
+    fn decode_events(&mut self, buf: &mut BytesMut) -> Vec<Event> {
+        let mut events = Vec::new();
+        loop {
+            match self.decoder.decode_eof(buf) {
+                Ok(Some((next, _))) => {
+                    events.extend(next.into_iter());
+                }
+                Ok(None) => break,
+                Err(error) => {
+                    // Error is logged by `crate::codecs::Decoder`, no further
+                    // handling is needed here.
+                    if !error.can_continue() {
+                        break;
+                    }
+                    break;
+                }
+            }
+        }
+        events
+    }
+
+    /// Enriches events with source_type, timestamp
+    fn enrich_events(&self, events: &mut Vec<Event>) {
+        for event in events {
+            match event {
+                Event::Log(ref mut log) => {
+                    log.try_insert(log_schema().source_type_key(), Bytes::from(NAME));
+                    log.try_insert(log_schema().timestamp_key(), Utc::now());
+                }
+                Event::Metric(ref mut metric) => {
+                    metric.insert_tag(log_schema().source_type_key().to_string(), NAME.to_string());
```

Comment:
Should this also set the metric timestamp if it's not present, as hinted to by the description?

Reply:
Gooood question. I based this function off of multiple other functions in different sources, but I recall not seeing timestamp being added for metrics, only for logs. I wasn't sure if that was intentional, but I can certainly try to add that logic here.

Reply:
Yes, it was something that occurred to me when reading this. I actually suspect the answer is "no" on further consideration, given issues we've had with users wanting metrics with no timestamp.